Automate the qualifications checks

robbiepaul opened the pull request:
<!-- Do you need to update CHANGELOG.md? -->
> So that service operators can perform claim checking faster and more reliably, we want to automate the qualifications checks (aka the DQT check) in the new task-based checking process.

## Screenshots

<img width="1101" alt="Screenshot 2020-03-18 at 10 55 17" src="https://user-images.githubusercontent.com/2804149/76953729-1d91bf80-6907-11ea-86eb-6ad0959671f1.png">
<img width="1110" alt="Screenshot 2020-03-18 at 10 55 28" src="https://user-images.githubusercontent.com/2804149/76953736-1ff41980-6907-11ea-9d47-18fe53e9b42d.png">
<img width="1142" alt="Screenshot 2020-03-18 at 10 55 49" src="https://user-images.githubusercontent.com/2804149/76953740-208cb000-6907-11ea-805c-582e331f0e2d.png">
<img width="1112" alt="Screenshot 2020-03-18 at 10 56 10" src="https://user-images.githubusercontent.com/2804149/76953745-21254680-6907-11ea-90d5-35b7e85f294d.png">


Sample CSV to use for testing, some of these will be eligible, some won't:

```csv
dfeta text1,dfeta text2,dfeta trn,fullname,birthdate,dfeta ninumber,dfeta qtsdate,dfeta he hesubject1idname,dfeta he hesubject2idname,dfeta he hesubject3idname,HESubject1Value,HESubject2Value,HESubject3Value,dfeta subject1idname,dfeta subject2idname,dfeta subject3idname,ITTSub1Value,ITTSub2Value,ITTSub3Value
1234567,JADA3SHJ,1234567,Fred Smith,17/3/2000,QQ123456C,23/8/2014,Politics,,,L200,,,Mathematics,,,G100,,
1234567,8N4EVBQN,1234567,Fred Smith,17/3/2000,QQ123456C,26/6/2015,,,,,,,General Primary,,,X120,,
1234567,9XJD85U5,1234567,Fred Smith,17/3/2000,QQ123456C,17/7/2015,,,,,,,Physics,,,F300,,
1234567,KJPC383U,1234567,Fred Smith,17/3/2000,QQ123456C,11/7/2013,Psychology,,,C800,,,General Primary,,,X120,,
1234567,P4SHNZV9,1234567,Fred Smith,17/3/2000,QQ123456C,18/12/2014,Psychology,,,C800,,,Mathematics,,,G100,,
1234567,GU4P79XF,1234567,Fred Smith,17/3/2000,QQ123456C,26/6/2014,Development studies,Biology,,L800,C100,,Biology,,,C100,,
1234567,Y6YX72GP,1234567,Fred Smith,17/3/2000,QQ123456C,20/6/2017,Psychology,,,C800,,,Mathematics,,,G100,,
1234567,3NRMMS56,1234567,Fred Smith,17/3/2000,QQ123456C,16/1/2015,,,,,,,Mathematics,,,G100,,
1234567,BKLKH9VE,1234567,Fred Smith,17/3/2000,QQ123456C,,,,,,,,,,,,,
1234567,7HWHXB52,1234567,Fred Smith,17/3/2000,QQ123456C,13/6/2014,Earth Science,,,F640,,,Physics,,,F300,,
```

robbiepaul commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Automate the qualifications checks (AKA DQT)](https://trello.com/c/iIUI63cU/1379-automate-the-qualifications-checks-aka-dqt)

pezholio commented on spec/requests/admin_qualification_report_upload_spec.rb:
```diff
@@ -0,0 +1,58 @@
+        eligibility: build(:maths_and_physics_eligibility, :eligible))
+    end
+
+    context "when the data in CSV matches the data in the claim" do
```
Should we do a check for an invalid CSV either here or in the feature too?

tekin commented on app/controllers/admin/qualification_report_uploads_controller.rb:
```diff
@@ -0,0 +1,17 @@
+
+    def create
+      @dqt_report_consumer = AutomatedChecks::DQTReportConsumer.new(params[:file], admin_user)
+      if @dqt_report_consumer.ingest
```
If we're careful to avoid N+1s and how we read the CSV I think we can avoid the complexity of moving this to a background job. The action is idempotent and a fast response isn't essential given the frequency this is done and the fact that it's for internal users.

tekin commented on app/models/academic_year.rb:
```diff
@@ -32,12 +32,17 @@ class << self
-        new(Date.today.year - 1)
+    # Returns the academic year for a given date, based on September 1st being
+    # the start of the year.
+    def for(date)
```
👍🏽

robbiepaul commented on app/models/maths_and_physics/dqt_record.rb:
```diff
@@ -0,0 +1,62 @@
+    def initialize(data)
+      @qts_award_date = Date.parse(data.fetch("dfeta qtsdate"))
+      @itt_subject_codes = data.slice("ITTSub1Value", "ITTSub2Value", "ITTSub3Value").compact.values
+      @postgraduate_codes = data.slice("HESubject1Value", "HESubject2Value", "HESubject3Value").compact.values
```
Yeah I flip-flopped between having them as private methods and in the constructor. I ended up choosing constructor because I thought that surfacing any errors closer to where they are passed in would be better. I'm happy to change it back though

pezholio replied:
Fair enough 👍

robbiepaul commented on app/models/automated_checks/dqt_report_csv.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+    end
+
+    def parse_csv(file)
```
I've been trying to use the '[rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming))' when it comes to refactoring duplication. I think the next time we have a class like this we should refactor then?

pezholio replied:
Yeah, that's fair enough, especially considering we know we will have something coming pretty soon

tekin commented on app/models/automated_checks/dqt_report_csv.rb:
```diff
@@ -0,0 +1,50 @@
+require "csv"
+
+module AutomatedChecks
+  class DQTReportCsv
```
Also have you seen https://ruby-doc.org/stdlib-2.6.3/libdoc/csv/rdoc/CSV.html#HeaderConverters? It might be a good idea to use that functionality to covert the wieldly named headers to headers that are more easy to work with elsewhere in the codebase. e.g. we'd be able to call `row.qts_year` instead of `row["dfeta qtsdate"]`.

robbiepaul replied:
I originally had a header converter in this class, however when it came to replaying the missing columns back to the user we felt it was quite confusing for the user when giving the converted headers in the error message. They would look at the CSV and none of the headings would match.

I'll have a think, to see if we can convert the headings but preserve the actual heading names for the error message.

robbiepaul commented on db/data/20200318093820_update_manual_on_tasks.rb:
```diff
@@ -0,0 +1,3 @@
+# Run me with `rails runner db/data/20200318093820_update_manual_on_tasks.rb`
+
+Task.update_all(manual: true)
```
Good point, I'll change it 👍

tekin commented on app/views/admin/qualification_report_uploads/new.html.erb:
```diff
@@ -7,6 +7,13 @@
       <div class="govuk-form-group">
         <%= f.label 'file', "Upload a CSV file", class: "govuk-label" %>
+
+        <% @dqt_report_consumer&.errors&.each do |error| %>
```
The double safe-navigations look a bit stinky. Something like this or wrap it in an `if`?:
```suggestion
        <% @dqt_report_consumer && @dqt_report_consumer.errors.each do |error| %>
```

tekin commented on app/views/admin/tasks/_task_outcome.html.erb:
```diff
@@ -3,6 +3,10 @@
+    <% if task.manual %>
+      This task was performed by <%= user_details(task.created_by, include_line_break: false) %> on <%= l(task.created_at) %>
+    <% else %>
+      This task was performed by an automated check on <%= l(task.created_at) %>
```
I think we should still include the name of the user here so it's clear who did the automated check.

robbiepaul commented:
@tekin I think I've addressed all the comments except the one about additional content. I'll have a chat with Faye/Rich after lunch and get something together for it.

tekin commented on spec/models/maths_and_physics/dqt_record_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require "rails_helper"
+
+RSpec.describe MathsAndPhysics::DQTRecord do
+  let(:maths_jac_code) { "G100" }
```
I think we should increase the coverage for the different variations of valid JACs codes, right now we are only testing two variations, which feels a bit thin. How about we use a list of valid and list of invalid ones and then iterated through those in the specs below?

robbiepaul replied:
There's dozens of [JAC codes](https://www.hesa.ac.uk/support/documentation/jacs/jacs3-detailed) so I've added five examples of each: maths, physics and ineligible. The ones I've used should cover each of the Maths JAC code prefixes (G1, G2, G3, G9) as well as Physics (F3)

tekin replied:
👍

robbiepaul commented on spec/models/automated_checks/dqt_report_consumer_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    it "doesn’t create a qualification task when the claim already has one" do
+      dqt_report_consumer.ingest
+
+      expect(claim_with_qualification_task.tasks.find_by(name: "qualifications")).to eql(existing_qualification_task)
```
Yeah I'm not sure, some major changes to the code would have to happen for it to fail. I guess it was more for documenting that the automated checks don't override tasks that have already been completed.

robbiepaul commented on spec/models/automated_checks/dqt_report_csv_spec.rb:
```diff
@@ -0,0 +1,85 @@
+      CSV
+    end
+
+    it "fails and populates its errors" do
```
I think I had meant to write `it fails to parse and` but the "fails to parse" bit is inconsequential so I've changed it to just be `it populates its errors`

tekin replied:
👍

robbiepaul commented:
@tekin Thanks for the re-review. I think I've addressed all the comments now.

I got some input from Rich and Faye on the content for the upload form, I've pushed up a WIP commit (359346536337daa25b612dc1d04435dbacf472a6) for it as I still need to find out the correct email address that DQT report requests get sent to.

tekin commented on spec/models/claim_spec.rb:
```diff
@@ -459,6 +459,18 @@
+    let!(:claim_with_employment_task) { create(:claim, :submitted, tasks: [build(:task, name: "employment")]) }
+    let!(:claim_with_qualification_task) { create(:claim, :submitted, tasks: [build(:task, name: "qualifications")]) }
+    let!(:claim_with_no_tasks) { create(:claim, :submitted, tasks: []) }
+    let!(:claim_with_decision) { create(:claim, :approved) }
```
This claim needs a task still:
```suggestion
    let!(:claim_with_decision) { create(:claim, :approved, tasks: [build(:task, name: "employment")]) }
```

tekin commented on spec/models/maths_and_physics/dqt_record_spec.rb:
```diff
@@ -0,0 +1,54 @@
+  MATHS_JAC_CODES = %w[G100 G290 G310 G320 G900]
+  PHYSICS_JAC_CODES = %w[F300 F310 F320 F321 F331]
+  ELIGIBLE_JAC_CODES = MATHS_JAC_CODES + PHYSICS_JAC_CODES
+  NON_ELIGIBLE_JAC_CODES = %w[X100 L800 F100 C700 R100]
```
Let's rename these to make it clearer that these are just examples, and not a definitive list:
```suggestion
  EXAMPLE_ELIGIBLE_JAC_CODES = MATHS_JAC_CODES + PHYSICS_JAC_CODES
  EXAMPLE_NON_ELIGIBLE_JAC_CODES = %w[X100 L800 F100 C700 R100]
```
